packages/astro/src/core/create-vite.ts

```diff
@@ -99,3 +106,85 @@ export async function createVite(inlineConfig: ViteConfigWithSSR, { astroConfig,
 	viteConfig = vite.mergeConfig(viteConfig, inlineConfig); // merge in inline Vite config
 	return viteConfig;
 }
+
+// Scans `projectRoot` for third-party Astro packages that could export an `.astro` file
+// `.astro` files need to be built by Vite, so these should use `noExternal`
+async function getAstroPackages({ projectRoot }: AstroConfig): Promise<string[]> {
+	const pkgUrl = new URL('./package.json', projectRoot);
+	const pkgPath = fileURLToPath(pkgUrl);
+	if (!fs.existsSync(pkgPath)) return [];
+
+	const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
+
+	const deps = [...Object.keys(pkg.dependencies || {}), ...Object.keys(pkg.devDependencies || {})];
+
+	return deps.filter((dep) => {
+		// Attempt: package is common and not Astro. ❌ Skip these for perf
+		if (isCommonNotAstro(dep)) return false;
+		// Attempt: package is named `astro-something`. ✅ Likely a community package
+		if (/^astro\-/.test(dep)) return true;
+		const depPkgUrl = new URL(`./node_modules/${dep}/package.json`, projectRoot);
+		const depPkgPath = fileURLToPath(depPkgUrl);
+		if (!fs.existsSync(depPkgPath)) return false;
+
+		const { dependencies = {}, peerDependencies = {}, keywords = [] } = JSON.parse(fs.readFileSync(depPkgPath, 'utf-8'));
+		// Attempt: package relies on `astro`. ✅ Definitely an Astro package
+		if (peerDependencies.astro || dependencies.astro) return true;
+		// Attempt: package is tagged with `astro` or `astro-component`. ✅ Likely a community package
+		if (keywords.includes('astro') || keywords.includes('astro-component')) return true;
+		return false;
```

Comment:
This is pretty well commented, but to explain the rationale:
1. Common packages will be encountered often. Don't need to scan these.
2. `astro-____` packages are probably community packages. Don't need to scan these.
3. Package has `astro` in `peerDependencies` (preferred) or `dependencies` (ok). It's an Astro package.
4. Package has `keywords` with `"astro"` or `"astro-component"`. We recommend these in the docs. It's probably an Astro package.
5. Anything else should be handled by Vite.